### ISSUE TITLE
some question about LabelDataset?

### ISSUE THREAD

**Issue by hquzhuguofeng** [soumyac1999/hyperbolic-label-emb-for-hmc#3]
Dear author:
         i have some question about the process of generating of the tensor(label_embs) that we use in calculate the loss.
N.shape = [n_labels, n_labels, n_labels]
and when we get the label_embs in batch in next(dataloader), why idx have relation with t,h
`    def __getitem__(self, idx):
          if idx >= self.len:
            raise StopIteration
          t = idx//self.items
          h = idx%self.items
          negs = np.arange(self.items)[self.N[t][h] == 1.0]
          negs = negs.repeat(self.nnegs)
          np.random.shuffle(negs)
          return torch.tensor([t, h, *negs[:self.nnegs]])
`

in my point of view t&h should relation with the content[idx]['label'] right? and you idea is idx relation with num_of_labels if i was wrong please tell me why? 
and another question is why the len of LabelDataset is num_of_labels*num_of_labels , not the len of dataset? so when we get the 
__getitem__ of the TextLabelDataset , 

`   def __getitem__(self, idx):
          text_data = self.text_dataset[idx%len(self.text_dataset)]
          label_data = self.label_dataset[idx%len(self.label_dataset)]
        return *self.text_dataset[idx%len(self.text_dataset)], self.label_dataset[idx%len(self.label_dataset)]
`

text_data and label_data may not consistent. if i was wrong please tell me why? 
I am looking forward to your rely, thanks a lot~

**Comment by soumyac1999** [comment]
Thanks for your questions.

Let me begin by clarifying that all sentences and their corresponding labels are stored in `TextDataset` objects, so when the `__getitem__` is called, a tokenized sentence and its corresponding multi-hot encoded label is returned. Thus, the text and labels are always consistent when calculating the first loss term (eq. 3 in the paper).

The purpose of the `LabelDataset` is to generate samples for the computation of the second loss term (eq. 4 in the paper). Note that this computation is independent of the first loss term, and thus different labels (from the same label set) can be present in the samples used in a single parameter update step.

Going back to what LabelDataset does, the matrix `N` corresponds to the matrix $\mathcal{L}(l, l')$ in the paper. When `__getitem__` is called a $l$ and $l'$ is selected (t and h in code) and `nneg` number of negative labels are sampled from `N`. (A fixed number `nneg` is sampled instead of using the entire $\mathcal{L}(l, l')$ to allow for vectorization.) The `LabelLoss` function in `main.py` might also help clarify how the samples from LabelDataset are used in the loss computation (this function is based on a similar one [here](https://github.com/facebookresearch/poincare-embeddings/blob/9aecb8634abbbc23338c2b5e5bb61ca517581bef/model.py#L114)).

Hope this helps answer your question.

**Comment by hquzhuguofeng** [comment]
thank you for your reply，now I fully understand your intelligent idea. Thanks a lot for your sharing~
